### PR TITLE
로그클래스와 적용시킨 trainer 합쳐주세요ㅎㅅㅎ

### PR DESCRIPTION
**Pull request by skrudtn**
gitignore에 logs/ 추가하고 README에 tensorboard 실행법 추가햇습니다.

### CODE REVIEW

File: codes/hyungsun/src/logger.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            run = int(os.listdir(log_dir)[len(os.listdir(log_dir))-1])+1
+
+        self.log_dir = os.path.join(log_dir, '%d' % run)
+        ensure_dir(log_dir)
+        self.writer = tf.summary.FileWriter(self.log_dir)
+        self.loss_sum = 0
+        self.accuracy_sum = 0
+        self.data_loader_len = data_loader_len
+
+    def scalar_summary(self, tag, value, step):
```
Comment (poqw) [inline codes/hyungsun/src/logger.py]: 이름은 `log` 로 간단하게 하는게 나을 것 같고, 코드가 길지 않으니 밑에 histo_summary랑 합친 후에 아래 처럼 enum써서 switch 하는 건 어떤가요?
```
class Mode(Enum):
     SCALAR = 1
     HISTOGRAM = 2
```
찾아보니 [이런 게](https://docs.python.org/3/library/enum.html) 있던데 참고 해주세요~

File: codes/hyungsun/src/train.py [hunk 2]
```diff
@@ -59,11 +60,25 @@ def train(self, max_epoch):
                 loss = self.criterion(output, target)
                 loss.backward()
                 self.optimizer.step()
+
+                # Compute Accuracy
+                _, argmax = torch.max(output,1)
+                accuracy = (target == argmax.squeeze()).float().mean()
+                self.logger.add_accuracy_and_loss(accuracy, loss)
```
Comment (poqw) [inline codes/hyungsun/src/train.py]: 만약 어떤 모델 쓰느냐에 따라서 정확도 계산하는 방식이 달라지면 모델이 정확도 계산하는 로직을 들고 있는 게 좋을 것 같은데, 너무 복잡해질것 같으니 이 부분은 일단 함수로 따로 빼두는 것만 해주세용

File: codes/hyungsun/README.md [hunk 3]
```diff
@@ -6,3 +6,13 @@
 # How to train.
 
     $ python3 train.py
+
+
+
+# How to run Tensorboard
+
+```
+$ tensorboard --logdir="log-folder-path" [--port="port-number"]
```
Comment (poqw) [inline codes/hyungsun/README.md]: 트레인 시작하면 자동으로 텐서보드 로깅되게 할 수 있지 않을까요?
https://stackoverflow.com/a/44896171

File: codes/hyungsun/src/train.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 from config import *
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/train.py]: Done.

File: codes/hyungsun/src/train.py [hunk 5]
```diff
@@ -19,6 +19,7 @@ def __init__(self, model, data_loader, optimizer, criterion, is_eval):
         self.default_filename = self.prefix + str(int(time.time())) + ".pt"
         self.current_epoch = 0
         self.criterion = criterion
+        self.logger = Logger('./logs',len(self.data_loader))
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/train.py]: Done.

File: codes/hyungsun/src/train.py [hunk 6]
```diff
@@ -114,8 +129,8 @@ def train_rnn_imdb():
 
 
 def main():
-    train_rnn_imdb()
-
+    # train_rnn_imdb()
+    train_cnn()
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/train.py]: Done.

File: codes/hyungsun/src/train.py [hunk 7]
```diff
@@ -59,11 +60,25 @@ def train(self, max_epoch):
...
+                self.logger.add_accuracy_and_loss(accuracy, loss)
+
                 if batch_idx % 100 == 0:
                     print("Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}".format(
                         max_epoch, batch_idx * len(data), len(self.data_loader.dataset),
                         100. * batch_idx / len(self.data_loader), loss.item()))
                     self.save_checkpoint()
+
+            # ====== Logging ===== #
+
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/train.py]: Done.

File: codes/hyungsun/src/logger.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            tag = tag.replace('.', '/')
+            self.histo_summary(tag, value.data.cpu().numpy(), cur_epoch + 1)
+            self.histo_summary(tag + '/grad', value.grad.data.cpu().numpy(), cur_epoch + 1)
+
+    def add_accuracy_and_loss(self, accuracy, loss):
+        self.accuracy_sum += accuracy.item()
+        self.loss_sum += loss.item()
+
+
+def ensure_dir(file_path):
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/logger.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
+import tensorflow as tf
+import numpy as np
+import scipy.misc
+import os
+import errno
+from io import BytesIO
+
+class Logger(object):
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/logger.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        self.loss_sum = 0
+        self.accuracy_sum = 0
+        self.data_loader_len = data_loader_len
+
+    def scalar_summary(self, tag, value, step):
+        """Log a scalar variable."""
+        summary = tf.Summary(value=[tf.Summary.Value(tag=tag, simple_value=value)])
+        self.writer.add_summary(summary, step)
+
+    def image_summary(self, tag, images, step):
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/logger.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        self.accuracy_sum += accuracy.item()
+        self.loss_sum += loss.item()
+
+
+def ensure_dir(file_path):
+    try:
+        os.makedirs(file_path)
+    except OSError as e:
+        if e.errno != errno.EEXIST:
+            raise
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/logger.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        """Create a summary writer logging to log_dir."""
+        ensure_dir(log_dir)
+        if len(os.listdir(log_dir)) == 0:
+            run = 1
+        else:
+            run = int(os.listdir(log_dir)[len(os.listdir(log_dir))-1])+1
+
+        self.log_dir = os.path.join(log_dir, '%d' % run)
+        ensure_dir(log_dir)
+        self.writer = tf.summary.FileWriter(self.log_dir)
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: log folder format - "Model Name"\"YYYY-MM-DD"T"hh.mm.ss"

File: codes/hyungsun/src/logger.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+import os
+import errno
+from io import BytesIO
+
+class Logger(object):
+    def __init__(self, log_dir, data_loader_len):
+        """Create a summary writer logging to log_dir."""
+        ensure_dir(log_dir)
+        if len(os.listdir(log_dir)) == 0:
+            run = 1
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/logger.py (hunk 1 above)
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/train.py [hunk 14]
```diff
@@ -19,6 +19,7 @@ def __init__(self, model, data_loader, optimizer, criterion, is_eval):
         self.default_filename = self.prefix + str(int(time.time())) + ".pt"
         self.current_epoch = 0
         self.criterion = criterion
+        self.logger = Logger('./logs',len(self.data_loader))
         
     def save_checkpoint(self):
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/train.py]: TODO(skrdutn): Make checkpoint for log

File: codes/hyungsun/src/logger.py [hunk 15]
```diff
@@ -1,103 +1,124 @@
...
+            step: Number of times data is processed.
+            bins: bin 이 int 지정된 범위(default = 10)의 동일한 컨테이너 수를 정의한다.
+                bin 이 sequence 인 경우 오른쪽 edge 를 지정하여 non-uniform bin width 를 허용한다.
+
+                Example:
+                    [ 0.5, 1.1, 1.3, 2.2, 2.9, 2.99 ]의 숫자 배열이 있을 때, bins 를 3으로 둔다면
+                    0 - 1 bin: (0.5)
+                    1 - 2 bin: (1.1, 1.3)
+                    2 - 3 bin: (2.2, 2.9, 2.99)
+                    위와같은 3개의 bin 으로 만들어 진다.
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.

File: codes/hyungsun/src/logger.py [hunk 16]
```diff
@@ -1,103 +1,124 @@
...
+
+    def summary(self, mode, tag, values, step, bins=1000):
+        """Log a scalar variable and histogram of the tensor of values.
+
+        Args:
+            mode: Enum to determine mode.
+            tag: The name of value data.
+            values: Data to log.
+            step: Number of times data is processed.
+            bins: bin 이 int 지정된 범위(default = 10)의 동일한 컨테이너 수를 정의한다.
```
[the remark this answers is not in the document]
Comment (skrudtn) [reply codes/hyungsun/src/logger.py]: Done.
